### ISSUE TITLE
Verifier Error / Incorrect comparisons against result of len() function

### ISSUE THREAD

**Issue by ajor** [bpftrace/bpftrace#3308]
This script should print "true":
```
BEGIN
{
  @[1] = 1;
  @[2] = 2;

  $count = len(@);
  // Uncomment this line and it works
  //$count = 2;

  printf("count: %d\n", $count);

  if ($count > 1) {
    print("true");
  } else {
    print("false");
  }

  printf("count: %d\n", $count);

  clear(@);
  exit();
}
```

Output:
```
Attaching 1 probe...
count: 2
false
count: 2
```

We can see that `$count` holds the value 2 (as expected), but this value does not evaluate as greater than 1 for some reason.
If we use the literal `2` instead of the result of `len(@)`, then we get the expected result.

**Comment by viktormalik** [comment]
Hm, my output on latest master (unless I did some mistake):
```
# src/bpftrace test.bt
[...]
BPF program is too large. Processed 1000001 insn
processed 1000001 insns (limit 1000000) max_states_per_insn 65 total_states 75003 peak_states 25025 mark_read 5


ERROR: Loading BPF object(s) failed.
```

**Comment by ajor** [comment]
> Hm, my output on latest master (unless I did some mistake):
> 
> ```
> # src/bpftrace test.bt
> [...]
> BPF program is too large. Processed 1000001 insn
> processed 1000001 insns (limit 1000000) max_states_per_insn 65 total_states 75003 peak_states 25025 mark_read 5
> 
> 
> ERROR: Loading BPF object(s) failed.
> ```

That's odd - my bpftrace compiles the script down to 168 instructions and I don't think there are any loops in there (except the for-each-map-elem).

Things that could be relevant are that I'm on LLVM 17 and my kernel is 5.19 but with lots of newer BPF features backported.

Here's my full `--info`:
```
System
  OS: Linux 5.19.0-0_fbk11_hardened_11538_g61e82533119f #1 SMP Wed Oct 4 09:20:00 PDT 2023
  Arch: x86_64
Build
  version: v0.21.0-33-g32ca
  LLVM: 17.0.6
  unsafe probe: no
  bfd: yes
  liblldb (DWARF support): yes
  libsystemd (systemd notify support): no
Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: yes
  get_boot_ns: yes
  dpath: yes
  skboutput: yes
  get_tai_ns: yes
  get_func_ip: yes
  jiffies64: yes
  for_each_map_elem: yes
Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf: yes
  module btf: yes
  map batch: yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes
Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes
  ringbuf: yes
Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  kprobe_multi: yes
  uprobe_multi: no
  raw_tp_special: yes
  iter: yes
```

**Comment by viktormalik** [comment]
Strange, I'm using the same LLVM with pretty much the same features:
```
# bpftrace --info
System
  OS: Linux 6.8.8-200.fc39.x86_64 #1 SMP PREEMPT_DYNAMIC Sat Apr 27 17:42:13 UTC 2024
  Arch: x86_64

Build
  version: v0.21.0-33-g32ca
  LLVM: 17.0.6
  unsafe probe: no
  bfd: yes
  liblldb (DWARF support): yes
  libsystemd (systemd notify support): no

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: no
  get_boot_ns: yes
  dpath: yes
  skboutput: yes
  get_tai_ns: yes
  get_func_ip: yes
  jiffies64: yes
  for_each_map_elem: yes

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf: yes
  module btf: yes
  map batch: yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes
  ringbuf: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  kprobe_multi: yes
  uprobe_multi: yes
  raw_tp_special: yes
  iter: yes
```
By a brief look, it seems that the for-loop callback body is expanded and inlined way too many times.

**Comment by jordalgo** [comment]
I'm seeing the same thing as @viktormalik but I'm on llvm12

```
$ sudo ./src/bpftrace --info
System
  OS: Linux 6.4.3-0_fbk11_2563_g44ec95f7d7f4 #1 SMP Thu Apr 18 16:09:12 PDT 2024
  Arch: x86_64

Build
  version: v0.21.0-33-g32ca
  LLVM: 12.0.1
  unsafe probe: no
  bfd: yes
  liblldb (DWARF support): yes
  libsystemd (systemd notify support): no

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: yes
  get_boot_ns: yes
  dpath: yes
  skboutput: yes
  get_tai_ns: yes
  get_func_ip: yes
  jiffies64: yes
  for_each_map_elem: yes

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf: yes
  module btf: yes
  map batch: yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes
  ringbuf: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  kprobe_multi: yes
  uprobe_multi: yes
  raw_tp_special: yes
  iter: yes
```

**Comment by viktormalik** [comment]
I had a closer look at the issue that I'm observing.

From verifier log:
```
[...]
31: (85) call bpf_for_each_map_elem#164       ; R0_w=scalar() R7=0 R10=fp0 fp-24=0 fp-32=mmmmmmmm
[...]
from 31 to 169: frame1: R1=map_ptr(map=AT_,ks=8,vs=8) R2=map_key(map=AT_,ks=8,vs=8) R3=map_value(map=AT_,ks=8,vs=8) R4=fp[0]-24 R10=fp0 cb
169: frame1: R1=map_ptr(map=AT_,ks=8,vs=8) R2=map_key(map=AT_,ks=8,vs=8) R3=map_value(map=AT_,ks=8,vs=8) R4=fp[0]-24 R10=fp0 cb
; 
169: (79) r1 = *(u64 *)(r4 +0)        ; frame1: R1_w=0 R4=fp[0]-24 cb
170: (07) r1 += 1                     ; frame1: R1_w=1 cb
171: (7b) *(u64 *)(r4 +0) = r1        ; frame1: R1_w=1 R4=fp[0]-24 cb
172: (b7) r0 = 0                      ; frame1: R0_w=0 cb
173: (95) exit 
mark_precise: frame1: last_idx 173 first_idx 31 subseq_idx -1 
mark_precise: frame1: regs=r0 stack= before 172: (b7) r0 = 0
returning from callee:                
 frame1: R0_w=0 R1_w=1 R2=map_key(map=AT_,ks=8,vs=8) R3=map_value(map=AT_,ks=8,vs=8) R4=fp[0]-24 R10=fp0 cb
to caller at 31:                      
 R0=scalar() R1=map_ptr(map=AT_,ks=8,vs=8) R2=func() R3=fp-24 R4=0 R7=0 R10=fp0 fp-24_w=1 fp-32=mmmmmmmm

from 173 to 31: R0=scalar() R1=map_ptr(map=AT_,ks=8,vs=8) R2=func() R3=fp-24 R4=0 R7=0 R10=fp0 fp-24=1 fp-32=mmmmmmmm
;
31: (85) call bpf_for_each_map_elem#164       ; R0_w=scalar() R7=0 R10=fp0 fp-24=1 fp-32=mmmmmmmm
[...]
```
Insn 31 is the call to `bpf_for_each_map_elem`. It has a jump to insn 169 which is where the callback function starts. Then, the callback function is analysed, ending with insn 173 which jumps back to 31. The problem here is that the state changed, `fp-24=1` as opposed to `fp-24=0` in the first visit to insn 31. This should be the counter value which we store on the stack.

I'm not sure if this is our problem or a verifier bug.

@ajor could you perhaps attach the output of your run with `-d verifier` option? That could shed some light on this issue.

**Comment by ajor** [comment]
This gets weirder...

This is my output with `-d verifier`, bearing in mind that the script runs for me without verifier complaints normally:
```
Attaching 1 probe...
BPF verifier log for BEGIN_1:
--------------------------------------

ERROR: Unknown BPF object load failure. Try using the "-d libbpf" option to see the full loading log.
ERROR: Loading BPF object(s) failed
```

With just `-d libbpf` (no `-d verifier`) I get:
```
[DEBUG] libbpf: loaded kernel BTF from '/sys/kernel/btf/vmlinux'
[DEBUG] libbpf: loaded kernel BTF from '/sys/kernel/btf/vmlinux'
[DEBUG] libbpf: loading object 'd0e150-1940' from buffer
[DEBUG] libbpf: elf: section(2) .text, size 40, link 0, flags 6, type=1
[DEBUG] libbpf: sec '.text': found program 'map_len_cb' at insn offset 0 (0 bytes), code size 5 insns (40 bytes)
[DEBUG] libbpf: elf: section(3) s_BEGIN_1, size 1352, link 0, flags 6, type=1
[DEBUG] libbpf: sec 's_BEGIN_1': found program 'BEGIN_1' at insn offset 0 (0 bytes), code size 169 insns (1352 bytes)
[DEBUG] libbpf: elf: section(4) .rels_BEGIN_1, size 256, link 20, flags 40, type=9
[DEBUG] libbpf: elf: section(5) license, size 4, link 0, flags 3, type=1
[DEBUG] libbpf: license of d0e150-1940 is GPL
[DEBUG] libbpf: elf: section(6) .maps, size 80, link 0, flags 3, type=1
[DEBUG] libbpf: elf: section(13) .BTF, size 798, link 0, flags 0, type=1
[DEBUG] libbpf: elf: section(15) .BTF.ext, size 120, link 0, flags 0, type=1
[DEBUG] libbpf: elf: section(17) .eh_frame, size 80, link 0, flags 2, type=1
[INFO] libbpf: elf: skipping unrecognized data section(17) .eh_frame
[DEBUG] libbpf: elf: section(18) .rel.eh_frame, size 32, link 20, flags 40, type=9
[INFO] libbpf: elf: skipping relo section(18) .rel.eh_frame for section(17) .eh_frame
[DEBUG] libbpf: elf: section(20) .symtab, size 648, link 1, flags 0, type=2
[DEBUG] libbpf: looking for externs among 27 symbols...
[DEBUG] libbpf: collected 0 externs total
[DEBUG] libbpf: map 'AT_': at sec_idx 6, offset 0.
[DEBUG] libbpf: map 'AT_': found type = 1.
[DEBUG] libbpf: map 'AT_': found key [8], sz = 8.
[DEBUG] libbpf: map 'AT_': found value [8], sz = 8.
[DEBUG] libbpf: map 'AT_': found max_entries = 4096.
[DEBUG] libbpf: map 'ringbuf': at sec_idx 6, offset 32.
[DEBUG] libbpf: map 'ringbuf': found type = 27.
[DEBUG] libbpf: map 'ringbuf': found max_entries = 262144.
[DEBUG] libbpf: map 'event_loss_counter': at sec_idx 6, offset 48.
[DEBUG] libbpf: map 'event_loss_counter': found type = 2.
[DEBUG] libbpf: map 'event_loss_counter': found key [20], sz = 4.
[DEBUG] libbpf: map 'event_loss_counter': found value [8], sz = 8.
[DEBUG] libbpf: map 'event_loss_counter': found max_entries = 1.
[DEBUG] libbpf: prog 'BEGIN_1': unrecognized ELF section name 's_BEGIN_1'
[DEBUG] libbpf: sec '.rels_BEGIN_1': collecting relocation for section(3) 's_BEGIN_1'
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #0: insn #8 against 'AT_'
[DEBUG] libbpf: prog 'BEGIN_1': found map 0 (AT_, sec 6, off 0) for insn #8
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #1: insn #19 against 'AT_'
[DEBUG] libbpf: prog 'BEGIN_1': found map 0 (AT_, sec 6, off 0) for insn #19
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #2: insn #26 against 'AT_'
[DEBUG] libbpf: prog 'BEGIN_1': found map 0 (AT_, sec 6, off 0) for insn #26
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #3: insn #28 against '.text'
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #4: insn #37 against 'ringbuf'
[DEBUG] libbpf: prog 'BEGIN_1': found map 1 (ringbuf, sec 6, off 32) for insn #37
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #5: insn #46 against 'event_loss_counter'
[DEBUG] libbpf: prog 'BEGIN_1': found map 2 (event_loss_counter, sec 6, off 48) for insn #46
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #6: insn #62 against 'ringbuf'
[DEBUG] libbpf: prog 'BEGIN_1': found map 1 (ringbuf, sec 6, off 32) for insn #62
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #7: insn #72 against 'event_loss_counter'
[DEBUG] libbpf: prog 'BEGIN_1': found map 2 (event_loss_counter, sec 6, off 48) for insn #72
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #8: insn #89 against 'ringbuf'
[DEBUG] libbpf: prog 'BEGIN_1': found map 1 (ringbuf, sec 6, off 32) for insn #89
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #9: insn #101 against 'ringbuf'
[DEBUG] libbpf: prog 'BEGIN_1': found map 1 (ringbuf, sec 6, off 32) for insn #101
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #10: insn #111 against 'event_loss_counter'
[DEBUG] libbpf: prog 'BEGIN_1': found map 2 (event_loss_counter, sec 6, off 48) for insn #111
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #11: insn #122 against 'ringbuf'
[DEBUG] libbpf: prog 'BEGIN_1': found map 1 (ringbuf, sec 6, off 32) for insn #122
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #12: insn #133 against 'event_loss_counter'
[DEBUG] libbpf: prog 'BEGIN_1': found map 2 (event_loss_counter, sec 6, off 48) for insn #133
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #13: insn #143 against 'ringbuf'
[DEBUG] libbpf: prog 'BEGIN_1': found map 1 (ringbuf, sec 6, off 32) for insn #143
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #14: insn #153 against 'event_loss_counter'
[DEBUG] libbpf: prog 'BEGIN_1': found map 2 (event_loss_counter, sec 6, off 48) for insn #153
[DEBUG] libbpf: sec '.rels_BEGIN_1': relo #15: insn #164 against 'event_loss_counter'
[DEBUG] libbpf: prog 'BEGIN_1': found map 2 (event_loss_counter, sec 6, off 48) for insn #164
Attaching 1 probe...
[DEBUG] libbpf: object 'd0e150-1940': failed (-22) to create BPF token from '/sys/fs/bpf', skipping optional step...
[DEBUG] libbpf: prog 'BEGIN_1': added 5 insns from sub-prog 'map_len_cb'
[DEBUG] libbpf: prog 'BEGIN_1': insn #28 relocated, imm 140 points to subprog 'map_len_cb' (now at 169 offset)
[DEBUG] libbpf: map 'AT_': created successfully, fd=3
[DEBUG] libbpf: map 'ringbuf': created successfully, fd=4
[DEBUG] libbpf: map 'event_loss_counter': created successfully, fd=5
```

With `-d libbpf -d verifier` I additionally see this at the end of the above log:
```
[WARN] libbpf: prog 'BEGIN_1': BPF program load failed: Invalid argument
[WARN] libbpf: prog 'BEGIN_1': failed to load: -22
[WARN] libbpf: failed to load object 'd0e100-1940
```

**Comment by ZhangYet** [comment]
same problem as `BPF program is too large. Processed 1000001 insn`

The kernel version is our company's internal version ,but we didn't modify the verifier.c.
```
System
  OS: Linux 6.6.0-47.001-rc1-xxxxxx
  Arch: x86_64

Build
  version: v0.21.0-153-g6782
  LLVM: 18.1.3
  unsafe probe: no
  bfd: yes
  liblldb (DWARF support): yes
  libsystemd (systemd notify support): no


  skboutput: yes
  get_tai_ns: yes
  get_func_ip: yes
  jiffies64: yes
  for_each_map_elem: yes

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf: yes
  module btf: yes
  map batch: yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes
  ringbuf: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  kprobe_multi: no
  uprobe_multi: yes
  raw_tp_special: yes
  iter: yes
```

I have traced the verifier code:

```c
static int do_check(struct bpf_verifier_env *env)
...
				if (state->curframe) {
					/* exit from nested function */
					err = prepare_func_exit(env, &env->insn_idx);
					if (err)
						return err;
					do_print_state = true;
					continue;
				}
...				
```

`prepare_func_exit` prints the `returning from callee ......` log, and it didn't return an error so the code continue, but somehow the `state->curframe` hadn't changed, so it went into a dead loop.

**Comment by ajor** [comment]
FWIW I'm now also seeing the "BPF program is too large. Processed 1000001 insn" verifier error after upgrading from that 5.19 kernel to 6.9.

**Comment by ajor** [comment]
Sounds like a potential verifier bug then?

Minimal repro:
```
BEGIN
{
  @[1] = 1;

  if (len(@) > 1) {
    @res = 1;
  }
}
```

Objdump of generated ELF file:
```
Disassembly of section .text:

0000000000000000 <map_len_cb>:
       0:       79 41 00 00 00 00 00 00 r1 = *(u64 *)(r4 + 0x0)
       1:       07 01 00 00 01 00 00 00 r1 += 0x1
       2:       7b 14 00 00 00 00 00 00 *(u64 *)(r4 + 0x0) = r1
       3:       b7 00 00 00 00 00 00 00 r0 = 0x0
       4:       95 00 00 00 00 00 00 00 exit

Disassembly of section s_BEGIN_1:

0000000000000000 <BEGIN_1>:
       0:       b7 06 00 00 01 00 00 00 r6 = 0x1
       1:       7b 6a f0 ff 00 00 00 00 *(u64 *)(r10 - 0x10) = r6
       2:       7b 6a f8 ff 00 00 00 00 *(u64 *)(r10 - 0x8) = r6
       3:       bf a2 00 00 00 00 00 00 r2 = r10
       4:       07 02 00 00 f0 ff ff ff r2 += -0x10
       5:       bf a3 00 00 00 00 00 00 r3 = r10
       6:       07 03 00 00 f8 ff ff ff r3 += -0x8
       7:       b7 07 00 00 00 00 00 00 r7 = 0x0
       8:       18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r1 = 0x0 ll
      10:       b7 04 00 00 00 00 00 00 r4 = 0x0
      11:       85 00 00 00 02 00 00 00 call 0x2
      12:       7b 7a f8 ff 00 00 00 00 *(u64 *)(r10 - 0x8) = r7
      13:       bf a3 00 00 00 00 00 00 r3 = r10
      14:       07 03 00 00 f8 ff ff ff r3 += -0x8
      15:       18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r1 = 0x0 ll
      17:       18 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r2 = 0x0 ll
      19:       b7 04 00 00 00 00 00 00 r4 = 0x0
      20:       85 00 00 00 a4 00 00 00 call 0xa4
      21:       79 a1 f8 ff 00 00 00 00 r1 = *(u64 *)(r10 - 0x8)
      22:       b7 02 00 00 02 00 00 00 r2 = 0x2
      23:       6d 12 0a 00 00 00 00 00 if r2 s> r1 goto +0xa <LBB0_2>
      24:       7b 7a f0 ff 00 00 00 00 *(u64 *)(r10 - 0x10) = r7
      25:       7b 6a f8 ff 00 00 00 00 *(u64 *)(r10 - 0x8) = r6
      26:       bf a2 00 00 00 00 00 00 r2 = r10
      27:       07 02 00 00 f0 ff ff ff r2 += -0x10
      28:       bf a3 00 00 00 00 00 00 r3 = r10
      29:       07 03 00 00 f8 ff ff ff r3 += -0x8
      30:       18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r1 = 0x0 ll
      32:       b7 04 00 00 00 00 00 00 r4 = 0x0
      33:       85 00 00 00 02 00 00 00 call 0x2

0000000000000110 <LBB0_2>:
      34:       b7 00 00 00 00 00 00 00 r0 = 0x0
      35:       95 00 00 00 00 00 00 00 exit
```

The verifier log is 15k lines log, but can be obtained with `bpftrace script.bt -d verifier`.

**Comment by 4ast** [comment]
The verifier cannot complete the loop, since there is precise check after the loop 'if r2 s> r1'. It makes fp-8 in the caller stack frame precise and the widening logic doesn't happen and all loop iterations look different, so it never converges.
I don't think we can do much about it.
It's best to post such issues at bpf @ vger to have a thread there and hopefully somebody can propose a workaround.

The workaround:
```
volatile u64 glob_var;

s_BEGIN_1(..)
{
  u64 ctx = 0;

  bpf_for_each_map_elem(... &ctx);
  glob_var = ctx;
  if (glob_var > 1)
    ...
}
```
in other words obfuscate loop callback ctx data via global variable.

**Comment by ZhangYet** [comment]
As the verifier log for the minimal repro:

```
returning from callee:
 frame1: R0_w=0 R1_w=0xc820 R2=map_key(map=AT_,ks=8,vs=8) R3=map_value(map=AT_,ks=8,vs=8) R4=fp[0]-8 R10=fp0 cb
to caller at 21:
 R0=scalar() R1=map_ptr(map=AT_,ks=8,vs=8) R2=func() R3=fp-8 R4=0 R6=1 R7=0 R10=fp0 fp-8_w=0xc820 fp-16=mmmmmmmm
```
It seems that the `for_map_each_elem` keeps calling the cb function but cannot break. Cannot understand why.

@ajor @4ast

**Comment by mannkafai** [comment]
Store the `len()` value in the map, that works fine in `6.8.0-45`. Seems to be a workaround.
 
```
BEGIN
{
  @[1] = 1;
  @[2] = 2;

  $count = len(@);
  @glob["count"] = $count;

  // Uncomment this line and it works
  // $count = 2;

  printf("count: %d\n", $count);

  if (@glob["count"] > 1) {
    print("true");
  } else {
    print("false");
  }

  printf("count: %d\n", $count);

  clear(@glob);
  clear(@);
  exit();
}
```

Output:

```
Attaching 1 probe...
count: 2
true
count: 2
```
